src/mailmap.c

```diff
@@ -0,0 +1,335 @@
+			entry->real_name = entry_data;
+			entry_data += real_name.size + 1; /* advance past null from calloc */
+		}
+		if (real_email.size > 0) {
+			memcpy(entry_data, real_email.ptr, real_email.size);
+			entry->real_email = entry_data;
+			entry_data += real_email.size + 1;
+		}
+		if (replace_name.size > 0) {
+			memcpy(entry_data, replace_name.ptr, replace_name.size);
+			entry->replace_name = entry_data;
+			entry_data += replace_name.size + 1;
+		}
+		/* replace_email is always non-null */
+		memcpy(entry_data, replace_email.ptr, replace_email.size);
+		entry->replace_email = entry_data;
+
+		error = git_vector_insert(&mm->entries, entry);
+		if (error < 0)
+			goto cleanup;
+		entry = NULL;
+	}
+
+	/* fill in *out, and make sure we don't free our mailmap */
+	*out = mm;
+	mm = NULL;
+
+cleanup:
+	git__free(entry);
+	git_mailmap_free(mm);
+
+	/* We never allocate data in these buffers, but better safe than sorry */
+	git_buf_free(&real_name);
+	git_buf_free(&real_email);
+	git_buf_free(&replace_name);
+	git_buf_free(&replace_email);
+	return error;
+}
+
+void git_mailmap_free(git_mailmap *mailmap)
+{
+	if (!mailmap)
+		return;
+
+	git_vector_free_deep(&mailmap->entries);
+	git__free(mailmap);
+}
+
+void git_mailmap_resolve(
+	const char **name_out, const char **email_out,
+	const git_mailmap *mailmap,
+	const char *name, const char *email)
+{
+	const git_mailmap_entry *entry = NULL;
+	assert(name && email);
+
+	*name_out = name;
+	*email_out = email;
+
+	if (!mailmap)
+		return;
+
+	entry = git_mailmap_entry_lookup(mailmap, name, email);
+	if (entry) {
+		if (entry->real_name)
+			*name_out = entry->real_name;
+		if (entry->real_email)
+			*email_out = entry->real_email;
+	}
+}
+
+const git_mailmap_entry *git_mailmap_entry_lookup(
+	const git_mailmap *mailmap, const char *name, const char *email)
+{
+	size_t i;
+	git_mailmap_entry *entry;
+	assert(name && email);
+
+	if (!mailmap)
+		return NULL;
+
+	git_vector_foreach(&mailmap->entries, i, entry) {
+		if (!git__strcmp(email, entry->replace_email) &&
+		    (!entry->replace_name || !git__strcmp(name, entry->replace_name))) {
+			return entry;
+		}
+	}
+
+	return NULL;
+}
+
+const git_mailmap_entry *git_mailmap_entry_byindex(
+	const git_mailmap *mailmap, size_t idx)
+{
+	if (mailmap)
+		return git_vector_get(&mailmap->entries, idx);
+	return NULL;
+}
+
+size_t git_mailmap_entry_count(const git_mailmap *mailmap)
+{
+	if (mailmap)
+		return git_vector_length(&mailmap->entries);
+	return 0;
+}
+
+static int mailmap_from_bare_repo(git_mailmap **mailmap, git_repository *repo)
+{
+	git_reference *head = NULL;
+	git_object *tree = NULL;
+	git_blob *blob = NULL;
+	git_buf content = GIT_BUF_INIT;
+	int error;
+
+	assert(git_repository_is_bare(repo));
+
+	/* In bare repositories, fall back to reading from HEAD's tree */
+	error = git_repository_head(&head, repo);
+	if (error < 0)
+		goto cleanup;
+
+	error = git_reference_peel(&tree, head, GIT_OBJ_TREE);
```

Comment:
No need for this. `git_object_lookup_bypath` will peel the treeish itself

Reply:
I believe that `head` is a `git_reference`, which I don't _think_ is a `git_object` which I can pass in and have peeled, though I may be wrong.

```
struct git_reference {
	git_refdb *db;
	git_ref_t type;

	union {
		git_oid oid;
		char *symbolic;
	} target;

	git_oid peel;
	char name[GIT_FLEX_ARRAY];
};
```
vs.
```
struct git_tree {
	git_object object;
	git_odb_object *odb_obj;
	git_array_t(git_tree_entry) entries;
};
```

Reply:
True